server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,874 @@
+        } else {
+            createSecondaryRepo(rollingRepositoryInitialRevision);
+        }
+    }
+
+    private void promoteSecondaryRepo(InternalRepository secondaryRepo,
+                                      Revision rollingRepositoryInitialRevision) {
+        writeLock();
+        try {
+            logger.info("Promoting the secondary repository in {}/{}.", parent.name(), originalRepoName);
+            repoMetadata.setPrimaryRepoDir(secondaryRepo.jGitRepo().getDirectory());
+            final InternalRepository primaryRepo = this.primaryRepo;
+            this.primaryRepo = secondaryRepo;
+            this.secondaryRepo = null;
+            repositoryWorker.execute(() -> {
+                closeInternalRepository(primaryRepo);
+                final File repoDir = primaryRepo.repoDir();
+                final Path path = repoDir.toPath();
+                final Path newPath = path.resolveSibling(path.getFileName().toString() + SUFFIX_REMOVED);
+                try {
+                    Files.move(path, newPath);
+                } catch (IOException e) {
+                    logger.warn("Failed to mark the old primary repository: {} to {}", repoDir, newPath);
+                }
+            });
+            logger.info("Promotion is done for {}/{}. {} is now the primary.",
+                        parent.name(), originalRepoName, secondaryRepo.repoDir());
+        } finally {
+            writeUnLock();
+        }
+
+        createSecondaryRepo(rollingRepositoryInitialRevision);
+    }
+
+    private void createSecondaryRepo(Revision rollingRepositoryInitialRevision) {
+        InternalRepository secondaryRepo = null;
+        try {
+            logger.info("Creating the secondary repository in {}/{}. head revision: {}.",
+                        parent.name(), originalRepoName, rollingRepositoryInitialRevision);
+            final List<Change<?>> changes = changes(rollingRepositoryInitialRevision);
+            final File secondaryRepoDir = repoMetadata.secondaryRepoDir();
+            secondaryRepo = InternalRepository.of(parent, originalRepoName, secondaryRepoDir,
+                                                  rollingRepositoryInitialRevision, creationTimeMillis,
+                                                  author, changes);
+            writeLock();
+            try {
+                final Revision headRevision = this.headRevision;
+                if (!rollingRepositoryInitialRevision.equals(headRevision)) {
+                    assert rollingRepositoryInitialRevision.major() < headRevision.major();
+                    // There were commits after the createRollingRepositoryCommand is created
+                    // so we should catch up.
+                    final RevisionRange revisionRange = new RevisionRange(
+                            rollingRepositoryInitialRevision.forward(1), headRevision);
+                    final List<Commit> commits = primaryRepo.listCommits(ALL_PATH, MAX_MAX_COMMITS,
```

Comment:
Q) I'm not sure how realistic/unrealistic the magic number 1000 is, but is it safe to say this won't be exceeded?
I'm asking because it seems like this implementation can break if `secondaryRepo.HEAD != primaryRepo.HEAD` (no problem if this isn't realistic)

I guess an alternative may be to allow partial `secondaryRepo` states.
This might also allow shorter write lock durations at the cost of additional complexity.

(or maybe we can just set an arbitrarily high value 🤔 )

Reply:
Or maybe we can also make `MAX_MAX_COMMITS` a variable and only log if `HEAD - INIT > MAX_MAX_COMMITS` 🤔